File: go/data_structures/linked_list.go
package main

import (
	"errors"
	"fmt"
)

var (
	ErrInvalidNodeIndex    = errors.New("node index invalid")
	ErrDeleteFromEmptyList = errors.New("can delete from empty list")
)

type Node struct {
	value string
	next  *Node
}

func NewNode(value string) *Node {
	return &Node{
		value: value,
	}
}

type LinkedList struct {
	head   *Node
	length int
}

func NewLinkedList() *LinkedList {
	return &LinkedList{}
}

func (l *LinkedList) isIndexValid(index int) bool {
	return index >= 0 && index < l.length
}

func (l *LinkedList) incrementLength() {
	l.length++
}

func (l *LinkedList) decrementLength() {
	l.length--
}

func (l *LinkedList) Empty() bool {
	return l.length == 0
}

func (l *LinkedList) Length() int {
	return l.length
}

func (l *LinkedList) Print() {
	node := l.head
	for node != nil {
		fmt.Printf("%s -> ", node.value)
		node = node.next
	}
	fmt.Println("None")
}

func (l *LinkedList) Tail() *Node {
	node := l.head
	index := l.length - 1
	for index > 0 {
		node = node.next
		index--
	}
	return node
}

func (l *LinkedList) GetNode(index int) (*Node, error) {
	if !l.isIndexValid(index) {
		return nil, ErrInvalidNodeIndex
	}

	if index == 0 {
		return l.head, nil
	}
	node := l.head
	for index > 0 {
		node = node.next
		index--
	}
	return node, nil
}

func (l *LinkedList) Append(node *Node) {
	if l.Empty() {
		l.head = node
	} else {
		tail := l.Tail()
		tail.next = node
	}
	l.incrementLength()
}

func (l *LinkedList) Insert(index int, node *Node) error {
	if !l.isIndexValid(index) {
		return ErrInvalidNodeIndex
	}

	if index == 0 {
		node.next = l.head
		l.head = node
		l.incrementLength()
		return nil
	}
	previousNode, err := l.GetNode(index - 1)
	if err != nil {
		return err
	}
	node.next = previousNode.next
	previousNode.next = node
	l.incrementLength()
	return nil
}

func (l *LinkedList) Delete(index int) error {
	previousNode, err := l.GetNode(index - 1)
	if err != nil {
		return err
	}
	previousNode.next = previousNode.next.next
	l.decrementLength()
	return nil
}

func (l *LinkedList) DeleteHead() error {
	if l.Empty() {
		return ErrDeleteFromEmptyList
	}
	l.head = l.head.next
	l.decrementLength()
	return nil

}

func (l *LinkedList) DeleteTail() error {
	if l.Empty() {
		return ErrDeleteFromEmptyList
	}
	err := l.Delete(l.length - 1)
	if err != nil {
		return err
	}
	return nil
}

func main() {
	linkedList := NewLinkedList()
	n4 := NewNode("forth")
	n3 := NewNode("third")
	n2 := NewNode("second")
	n1 := NewNode("first")

	err := linkedList.DeleteHead()
	if err != nil {
		fmt.Println(err)
	}

	linkedList.Append(n1)
	linkedList.Append(n3)
	linkedList.Append(n4)
	linkedList.Append(n2)

	linkedList.Print()
	fmt.Println(linkedList.Length())

	err = linkedList.Delete(5)
	if err != nil {
		fmt.Println(err)
	}
	err = linkedList.Delete(2)
	if err != nil {
		fmt.Println(err)
	}
	linkedList.Print()
	err = linkedList.DeleteHead()
	if err != nil {
		fmt.Println(err)
	}
	linkedList.Print()
	fmt.Println(linkedList.Length())
	err = linkedList.DeleteTail()
	if err != nil {
		fmt.Println(err)
	}

	linkedList.Print()
	fmt.Println(linkedList.Length())

	m4 := NewNode("m_forth")
	m3 := NewNode("m_third")
	m2 := NewNode("m_second")
	m1 := NewNode("m_first")

	linkedList.Append(m1)
	linkedList.Append(m4)

	linkedList.Print()
	fmt.Println(linkedList.Length())

	err = linkedList.Insert(0, m3)
	if err != nil {
		fmt.Println(err)
	}

	linkedList.Print()
	fmt.Println(linkedList.Length())

	err = linkedList.Insert(4, m2)
	if err != nil {
		fmt.Println(err)
	}

	err = linkedList.Insert(2, m2)
	if err != nil {
		fmt.Println(err)
	}

	linkedList.Print()
	fmt.Println(linkedList.Length())

}

File: python/src/data_structures/linked_list.py
from typing import Self
from unittest.util import _MAX_LENGTH


class Node:
    def __init__(self, value: str, _next: Self | None) -> None:
        self._value = value
        self._next = _next

    def clear_next(self) -> None:
        self._next = None

    @property
    def value(self) -> str:
        return self._value

    @property
    def next(self) -> Self | None:
        return self._next

    @next.setter
    def next(self, value: Self | None) -> None:
        self._next = value


class LinkedList:
    _head: Node

    def __init__(self) -> None:
        self._length = 0

    def __len__(self) -> int:
        return self._length

    def __iter__(self) -> Self:
        self._current = self._head
        return self

    def __next__(self) -> str:
        if self._current is None:
            raise StopIteration
        value = self._current.value
        self._current = self._current.next
        return value

    def __str__(self) -> str:
        result = []
        for i, v in enumerate(self):
            result.append(f"[{i}: {v}]")
        return " -> ".join(result)

    def get_node(self, index: int) -> Node | None:
        if index < 0 or index >= len(self):
            return None

        node = self._head
        for _ in range(index):
            if node is None:
                return None
            node = node.next
        return node

    def append(self, value: str) -> None:
        node = self.get_node(len(self) - 1)
        if node is None:
            self._head = Node(value, None)
        else:
            node.next = Node(value, None)
        self._length += 1

    def insert(self, index: int, value: str) -> None:
        if index < 0 or index > len(self):
            return

        if index == 0:
            self._head = Node(value, self._head)
        else:
            node = self.get_node(index - 1)
            if node is not None:
                node.next = Node(value, node.next)
        self._length += 1

    def remove(self, index: int) -> None:
        if index < 0 or index >= len(self):
            return

        if index == 0:
            if self._head is not None:
                self._head = self._head.next
        else:
            node = self.get_node(index - 1)
            if node is not None:
                next_node = node.next
                if next_node is not None:
                    node.next = next_node.next
        self._length -= 1


if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.append("first")
    linked_list.append("second")
    linked_list.append("third")
    print(linked_list)

    linked_list.insert(1, "1node")
    print(linked_list)

    linked_list.remove(1)
    print(linked_list)

    linked_list.insert(3, "3node")
    print(linked_list)

    linked_list.remove(3)
    print(linked_list)

    linked_list.insert(4, "4node")
    print(linked_list)

    linked_list.remove(4)
    print(linked_list)
